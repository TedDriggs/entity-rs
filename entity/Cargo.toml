[package]
name = "entity"
description = "Library that provides entity-like constructs"
categories = ["data-structures"]
version = "0.1.0"
authors = ["Chip Senkbeil <chip@senkbeil.org>"]
edition = "2018"
homepage = "https://github.com/chipsenkbeil/entity-rs"
repository = "https://github.com/chipsenkbeil/entity-rs"
readme = "README.md"
license = "MIT OR Apache-2.0"

[features]
default = ["inmemory_db", "sled_db"]
macros = ["entity_macros"]
inmemory_db = ["serde", "serde/rc", "typetag"]
sled_db = ["bincode", "serde", "sled", "typetag"]

[dependencies]
bincode = { version = "1.3.1", optional = true }
derive_more = { version = "0.99.11", default-features = false, features = ["as_ref", "as_mut", "constructor", "deref", "deref_mut", "display", "error", "from", "into", "try_into"] }
dyn-clone = "1.0.3"
paste = "1.0.3"
serde = { version = "1.0.117", features = ["derive"], optional = true }
sled = { version = "0.34.6", optional = true }
strum = { version = "0.19", features = ["derive"] }
typetag = { version = "0.1.6", optional = true }
entity_macros = { version = "0.1.0", path = "../entity_macros", optional = true }
