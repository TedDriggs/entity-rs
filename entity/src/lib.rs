mod any;
mod context;
mod field;
pub mod query;
mod value;

pub use any::AsAny;
pub use context::Context;
pub use field::{Field, FieldDefinition};
pub use query::{Query, QueryExt};
pub use value::{PrimitiveValue, PrimitiveValueType, Value, ValueType};

use derive_more::Constructor;
use std::collections::HashMap;

/// Represents a generic entity
pub trait Ent: AsAny {
    /// Represents the unique id associated with each entity instance
    fn id(&self) -> usize;

    /// Represents a unique type associated with the entity, used for
    /// lookups, indexing by type, and conversions
    fn r#type(&self) -> &'static str;

    /// Represents fields contained within the ent instance, guaranteeing
    /// that fields are unique by name
    fn fields(&self) -> Vec<&Field>;

    /// Retrieves the field with the provided name within the ent instance
    fn field(&self, name: &str) -> Option<&Field>;

    /// Retrieves the value for the field with the provided name within the
    /// ent instance
    fn field_value(&self, name: &str) -> Option<&Value> {
        self.field(name).map(|f| f.value())
    }
}

/// Represents an ent that uses an internal map to contain fields dynamically
#[derive(Constructor, Debug)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct MapEnt(usize, HashMap<String, Field>);

impl_as_any!(MapEnt);

impl MapEnt {
    /// Creates an empty map ent with the provided id
    pub fn empty(id: usize) -> Self {
        Self::new(id, HashMap::new())
    }
}

impl Default for MapEnt {
    /// Creates an empty map ent using 0 as the id
    fn default() -> Self {
        Self::empty(0)
    }
}

impl Ent for MapEnt {
    /// Represents the unique id associated with each entity instance
    ///
    /// ## Examples
    ///
    /// ```
    /// use entity::{Ent, MapEnt};
    ///
    /// let ent = MapEnt::empty(999);
    /// assert_eq!(ent.id(), 999);
    /// ```
    fn id(&self) -> usize {
        self.0
    }

    /// Represents a unique type associated with the entity, used for
    /// lookups, indexing by type, and conversions
    ///
    /// ## Examples
    ///
    /// ```
    /// use entity::{Ent, MapEnt};
    ///
    /// let ent = MapEnt::default();
    /// assert_eq!(ent.r#type(), "MapEnt");
    /// ```
    fn r#type(&self) -> &'static str {
        "MapEnt"
    }

    /// Represents fields contained within the ent instance
    ///
    /// ## Examples
    ///
    /// ```
    /// use entity::{Ent, MapEnt, Field};
    ///
    /// let fields = vec![
    ///     Field::new("field1", 123u8),
    ///     Field::new("field2", "some text"),
    /// ];
    /// let ent = MapEnt::new(0, fields.iter().map(|f| (f.name().to_string(), f.clone())).collect());
    ///
    /// let ent_fields = ent.fields();
    /// assert_eq!(ent_fields.len(), 2);
    /// assert!(ent_fields.contains(&&Field::new("field1", 123u8)));
    /// assert!(ent_fields.contains(&&Field::new("field2", "some text")));
    /// ```
    fn fields(&self) -> Vec<&Field> {
        self.1.values().collect()
    }

    /// Retrieves the field with the provided name within the ent instance
    ///
    /// ## Examples
    ///
    /// ```
    /// use entity::{Ent, MapEnt, Field, Value};
    ///
    /// let fields = vec![
    ///     Field::new("field1", 123u8),
    ///     Field::new("field2", "some text"),
    /// ];
    /// let ent = MapEnt::new(0, fields.iter().map(|f| (f.name().to_string(), f.clone())).collect());
    ///
    /// assert_eq!(ent.field("field1").unwrap().value(), &Value::from(123u8));
    /// assert_eq!(ent.field("field???"), None);
    /// ```
    fn field(&self, name: &str) -> Option<&Field> {
        self.1.get(name)
    }
}
