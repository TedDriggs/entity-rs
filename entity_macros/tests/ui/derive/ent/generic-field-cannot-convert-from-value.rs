use entity::{Database, Ent, Id, Value};

#[derive(Clone, Ent)]
struct TestEnt<T: Clone + Into<Value> + 'static> {
    #[ent(id)]
    id: Id,

    #[ent(database)]
    database: Option<Box<dyn Database>>,

    #[ent(created)]
    created: u64,

    #[ent(last_updated)]
    last_updated: u64,

    #[ent(field)]
    my_field: T,
}

fn main() {}
