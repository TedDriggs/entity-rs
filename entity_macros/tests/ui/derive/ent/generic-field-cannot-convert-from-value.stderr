error[E0271]: type mismatch resolving `<T as std::convert::TryFrom<entity::Value>>::Error == &str`
 --> $DIR/generic-field-cannot-convert-from-value.rs:3:17
  |
3 | #[derive(Clone, Ent)]
  |                 ^^^ expected enum `std::convert::Infallible`, found `&str`
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `T: std::convert::From<entity::Value>` is not satisfied
 --> $DIR/generic-field-cannot-convert-from-value.rs:3:17
  |
3 | #[derive(Clone, Ent)]
  |                 ^^^ the trait `std::convert::From<entity::Value>` is not implemented for `T`
  |
  = note: required because of the requirements on the impl of `std::convert::Into<T>` for `entity::Value`
  = note: required because of the requirements on the impl of `std::convert::TryFrom<entity::Value>` for `T`
  = note: required because of the requirements on the impl of `std::convert::TryInto<T>` for `entity::Value`
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
  |
4 | struct TestEnt<T: Clone + Into<Value> + 'static + std::convert::From<entity::Value>> {
  |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
